File: leetcode/code/39.combination-sum.py
def backtrack(
    solutions: list[list[int]],
    vec: list[int],
    sum: int,
    target: int,
    candidates: list[int],
    candidate_idx: int,
):
    if sum == target:
        solutions.append(vec)
        return
    if candidate_idx >= len(candidates) or sum + candidates[candidate_idx] > target:
        return

    value = candidates[candidate_idx]

    backtrack(solutions, vec, sum, target, candidates, candidate_idx + 1)
    while sum < target:
        sum += value
        if sum > target:
            break
        vec = vec[:]
        vec.append(value)
        backtrack(
            solutions,
            vec,
            sum,
            target,
            candidates,
            candidate_idx + 1,
        )


class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        solutions: list[list[int]] = []
        backtrack(solutions, [], 0, target, sorted(candidates), 0)
        return solutions

File: leetcode/code/39.combination-sum.rs
#[derive(Clone)]
struct Data {
    solution: Vec<i32>,
    sum: i32,
}

fn backtrack(data: &Data, target: i32, candidates: &[i32]) -> Vec<Data> {
    if data.sum == target {
        vec![data.clone()]
    } else if data.sum > target || candidates.is_empty() {
        vec![]
    } else {
        let needed = target - data.sum;
        let value = candidates[0];
        let mut data = data.clone();
        let mut output = vec![];

        output.extend_from_slice(&backtrack(&data, target, &candidates[1..]));
        for _ in 0..needed / value {
            data.solution.push(value);
            data.sum += value;
            output.extend_from_slice(&backtrack(&data, target, &candidates[1..]));
        }
        output
    }
}

impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        backtrack(
            &Data {
                solution: vec![],
                sum: 0,
            },
            target,
            &candidates,
        )
        .into_iter()
        .map(|v| v.solution)
        .collect()
    }
}
